enforcer/src/main/java/org/wso2/micro/gateway/enforcer/security/AuthFilter.java

```diff
@@ -46,15 +49,65 @@ public boolean handleRequest(RequestContext requestContext) {
                 if (authenticator.canAuthenticate(requestContext)) {
                     AuthenticationContext authenticate = authenticator.authenticate(requestContext);
                     if (authenticate.isAuthenticated()) {
+                        updateClusterHeaderAndCheckEnv(requestContext, authenticate);
                         return true;
                     }
                 }
             }
         } catch (APISecurityException e) {
-            requestContext.getProperties().put("code", e.getErrorCode());
+            //TODO: (VirajSalaka) provide the error code properly based on exception (401, 403, 429 etc)
+            requestContext.getProperties().put("code", "401");
             requestContext.getProperties().put("error_code", e.getErrorCode());
             requestContext.getProperties().put("error_description", e.getMessage());
         }
         return false;
     }
+
+    /**
+     * Update the cluster header based on the keyType and authenticate the token against its respective endpoint
+     * environment.
+     * 
+     * @param requestContext request Context 
+     * @param authContext authentication context
+     * @throws APISecurityException if the environment and 
+     */
+    private void updateClusterHeaderAndCheckEnv(RequestContext requestContext, AuthenticationContext authContext)
+            throws APISecurityException {
+
+        String keyType = authContext.getKeyType();
+        if (StringUtils.isEmpty(authContext.getKeyType())) {
+            keyType = "PRODUCTION";
+        } 
+
+        // Header needs to be set only if the relevant cluster is available for the resource and the key type is
+        // matched.
+        if (requestContext.isClusterHeaderEnabled()) {
+            if (keyType.equalsIgnoreCase("PRODUCTION") &&
+                    !StringUtils.isEmpty(requestContext.getProdClusterHeader())) {
+                requestContext.addResponseHeaders(AdapterConstants.CLUSTER_HEADER,
+                        requestContext.getProdClusterHeader());
+            } else if (keyType.equalsIgnoreCase("SANDBOX") &&
+                    !StringUtils.isEmpty(requestContext.getSandClusterHeader())) {
```

Comment:
Remove the redundant isEmpty check

Reply:
fixed via https://github.com/wso2/product-microgateway/pull/1525/commits/452e70df3c574001865304cc59dcf25ad30a79e9